### PR TITLE
Feat/use zetasql as front end

### PR DESCRIPTION
**Pull request by jingchen2222**
## Summary

We are going to use ZetaSQL as our new SQL front-end. The document is trying to figure out:

- **Motivation**: Why we want to refactor the front-end
- **Feasibility**: Whether ZetaSQL is a good option or not
- **Design**: How to integrate ZetaSQL into HybridSE
- **Adoption strategy**: What breaking changes it will bring to FEDB and SparkFE

## Links

- [Full Rendered Proposal](https://github.com/jingchen2222/rfcs/blob/feat/use-zetasql-as-front-end/hybridse/using-zetasql.md)

- [Discussion Thread]()

---

**Important: Do NOT comment on this PR. Please use the discussion thread linked above to provide feedback, as it provides branched discussions that are easier to follow. This also makes the edit history of the PR clearer.**

### CODE REVIEW

**Comment by imotai** [comment]
@jingchen2222 Please add  the following sections

* design of syntax error tips 
* design of  `functions not implemented ` error tips

all tips must include help links

File: hybridse/using-zetasql.md [hunk 1]
```diff
@@ -0,0 +1,294 @@
...
+
+### Last Join
+
+**Solution**:
+
+the `ORDER BY` option can be added into `opt_on_or_using_clause_list`  and add code to deal with it.
+
+### Window ROWS_RANGE
+
+- Make sure HybridSE `udf` and `udaf` works with zetasql - todo
```
Comment (imotai) [inline hybridse/using-zetasql.md]: how to add external udf to zetasql ?

File: hybridse/using-zetasql.md [hunk 2]
```diff
@@ -0,0 +1,294 @@
...
+--
+ERROR: Syntax error: Unexpected ";" [at 1:11]
+drop table;
+          ^
+==
+```
+
+
+
+#### All tips must be include help links
```
Comment (imotai) [inline hybridse/using-zetasql.md]: must be include -> must include

File: hybridse/using-zetasql.md [hunk 3]
```diff
@@ -0,0 +1,294 @@
...
+
+```SQL
+select from t
+--
+ERROR: Syntax error: SELECT list must not be empty [at 1:9]
+select  from t
+        ^
+If the tips above don't help, you can get more help from [slack channel](https://hybridsql-ws.slack.com/archives/C01R7LAF6AY)
+```
```
Comment (imotai) [inline hybridse/using-zetasql.md:646]: https://join.slack.com/t/hybridsql-ws/shared_invite/zt-ozu3llie-K~hn9Ss1GZcFW2~K_L5sMg

File: hybridse/using-zetasql.md (hunk 1 above)
Comment (jingchen2222) [reply hybridse/using-zetasql.md]: Status: todo.
We will to describe it in this section.

File: hybridse/using-zetasql.md [hunk 4]
```diff
@@ -0,0 +1,294 @@
...
+
+| os        | Clang                                           | build | test |      |
+| --------- | ----------------------------------------------- | ----- | ---- | ---- |
+| Mac:10.15 | Apple clang version 12.0.0 (clang-1200.0.32.21) | pass  | Pass |      |
+
+Mac building status: https://github.com/jingchen2222/zetasql/runs/2348401484?check_suite_focus=true
+
+## Install (WIP)
+
+We create a cmake sampel project integrated zetasql. Check [Discussion link](https://github.com/4paradigm/HybridSE/discussions/42) for more details.
```
Comment (aceforeverd) [inline hybridse/using-zetasql.md]: Wrote a summary about manage zetasql as cmake dependency.
https://github.com/4paradigm/HybridSE/discussions/42#discussioncomment-655446

Comment (jingchen2222) [reply hybridse/using-zetasql.md]: Great job.

File: hybridse/using-zetasql.md [hunk 5]
```diff
@@ -0,0 +1,575 @@
...
+We are going to use ZetaSQL as our new SQL front-end. The document is trying to figure out:
+
+- **Motivation**: Why we want to refactor the front-end
+- **Feasibility**: Whether ZetaSQL is a good option or not
+- **Design**: How to integrate ZetaSQL into HybridSE
+- **Adoption strategy**: What breaking changes it will bring to FEDB and SparkFE
+
+# Motivation
+
+HybridSE is LLVM-based SQL engine written in C++. It aims at supporting statndare ANSI SQL and some new SQL traits at the same time. However, as a newly engine, HybridSE only supports a limited set of SQL syntaxs currently and its syntax error message is unfriendly. For example, the syntax error message can't tell the location of wrong expression.
```
Comment (imotai) [inline hybridse/using-zetasql.md]: `statndare` typo

File: hybridse/using-zetasql.md [hunk 6]
```diff
@@ -0,0 +1,575 @@
...
+
+ZetaSQL is a SQL Analyzer Framework from Google. We are considering using it as HybridSE's SQL parser and analyzer for following reasons:
+
+- ZetaSQL has been used for BigQuery, Spanner and DataflowSQL by Google
+- ZetaSQL is a C++ SQL parser, which can be easily integrate into HybridSE
+- ZetaSQL supports standare ANSI SQL and provides clearly error messge. 
+
```
Comment (imotai) [inline hybridse/using-zetasql.md:26]: standard

**Comment by jingchen2222** [comment]
Update @imotai  @aceforeverd 
- Add Design for hybridse `CREATE PROCEDURE` with `CONST column_name column_type` parameters https://github.com/4paradigm/rfcs/pull/3/commits/6a5522035001a1deec9a61c0f4f49cb7e57a4f0a
- Add Test description: https://github.com/4paradigm/rfcs/pull/3/commits/63cd99dc05a1cb8064cf206cc32188a79a285e24

**Review by imotai — APPROVED** [review]
LGTM

**Review by aceforeverd — APPROVED** [review]

**Review by tobegit3hub — APPROVED** [review]
LGTM

**Pull request merged**